File: leetcode/382.链表随机节点/382.链表随机节点.go
package main

import "math/rand"

type ListNode struct {
	Val  int
	Next *ListNode
}

type Solution struct {
	head *ListNode
}

func Constructor(head *ListNode) Solution {
	return Solution{
		head: head,
	}
}

/** Returns a random node's value. */
func (this *Solution) GetRandom() int {
	count := 0
	res := 0
	cur := this.head
	for cur != nil {
		count += 1
		randNum := rand.Intn(count) + 1
		if randNum == count {
			res = cur.Val
		}
		cur = cur.Next
	}
	return res
}

/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(head);
 * param_1 := obj.GetRandom();
 */

File: leetcode/382.链表随机节点/382.链表随机节点.py
import random
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    """
    遍历一遍，抽取一个元素，保证每个节点获得的概率是一样的
    蓄水池算法
    每次只保留一个数，当遇到第 i 个数时，以 1/i的概率保留它，(i-1)/i的概率保留原来的数。
    """
    def __init__(self, head: Optional[ListNode]):
        self.head = head

    def getRandom(self) -> int:
        count = 0
        res = 0
        cur = self.head
        while cur:
            count += 1
            rand = random.randint(1, count)
            if rand == count:
                res = cur.val
            cur = cur.next
        return res
